### ISSUE TITLE
Customize OpMaker for MKL-DNN builds

### ISSUE THREAD

**Issue by jczaja** [PaddlePaddle/Paddle#17960]
Hi,

I want to solve following problem:
Activations of paddle (Grad ops) are using inputs: Out, Out@GRAD   while  mkl-dnn grad primitives  activations need: X  (input signal to fwd op)  . eg. MKL-DNN grad activation ops are expecting input of forward to be available.
So far we used SetBlob to have it transfered, but this is problematic for parallel executor.
Hence, please advice how can I modify paddle so that In activation grad ops (for mkl-dnn builds), X (input signal to fwd op) is available in grad op ?  In other words :
ExecutionContext::OperatorBase::inputs_  should have:
{["Out"] = vector<> , ["Out@GRAD"] = vector<>, **["X"]= vector<>**}

Please advice on how to do so if it is possible?

**Comment by jczaja** [comment]
@reyoung , @tensor-tang , @luotao1 Could you please advice on this issue?

**Comment by ghost** [comment]
@jczaja Could you please rewrite the description for this issue? We (Luo Tao, Leo, and me) all have difficulty to understand it. But I told Luo Tao what you are asking based on our previous discussion. Currently we use SetBlob to save info (e.g. pd/workspace_memory) in FWD OP and then use GetBlob to retrieved those saved info which is needed in BWD OP. For example, we save pooling pd in Pool OP and retrieve it in Pool Grad OP. 
However, this is not a good solution. The best way to pass info between FWD OP and BWD OP are through output tensors of FWD OP. This is because by design all input/output tensors of FWD OP will be passed to BWD OP as inputs.

Based on above description, could you rewrite your question again? And please add:
1. An example 
2. Why this issue is mandatory for "activation" op? why SetBlob/GetBlob way can't apply to it?

Thanks!

**Comment by jczaja** [comment]
Currently(unless something changed) for parallel executor we use std::thread_id to distinguish caching
for one ParallelExecutor thread from other ParallelExecutor thread. But there is once exception which
is when data is to be transfered from Forward Op to Grad Op then TID is **not** used. This is because
diffrent threads may execute Grad ops and Forward Ops.   For activation mkl-dnn ops flow is like this:

<img src="https://raw.githubusercontent.com/jczaja/Paddle/pictures/images/customize.svg?sanitize=true">

For other ops where MKL-DNN PD is transfered from Forward to Backward this is fine, but here (activation mkldnn) there may be a problem as without TID as part of Hash Key effectively
we are sharing X among Parallel Executor threads. Yes Tensor pointers will be different
but pointer is installed within MKL-DNN memory object which is common to parallel executor threads.
So threre is a race condition related to changing pointer to Tensor Data . At the end Backward may not get X data it expected. Here is picture of what I want to achive eg. connect X to activation grad via graph not cache:

<img src="https://raw.githubusercontent.com/jczaja/Paddle/pictures/images/customize_after.svg?sanitize=true">

To summarize, By storing X in Cache we are trying , blindly to replicate what should be done on graph level like. X should be transfered via graph the same way Out is connected from Forward to Backward.
So this issue is on  questions:
1) how Can We modify graph so that for MKL-DNN activation ops X is connected to corressponding grad mkl-dnn op ?
2) Is there any alternative eg. enforcing so that forward and backward op are executed by the same thread?

**Comment by ghost** [comment]
@jczaja I guess I understand your concern now. 
Answer to your question 1: All inputs/outputs vars of fwd OP (i.e. here "X" and "Out" of activation op) will be automatically set as input vars of bwd OP (i.e. activation_grad op).
In another word, the input vars of "activation_grad" op are: "X", "Out", "diff Y". It's original design of paddle already.
So I don't understand why you need pass x_data from fwd to bwd.

**Comment by jczaja** [comment]
@jianhang-liu activation_grad does **not** have "X" as input. Only: "Out" and "diff Y" are inputs because paddle naive grad implementation of activation does not need "X" it does need "Out" . MKL-DNN does need "X" to compute grad. Second image I put is description of what I want it to be . The first one is how it works right now

**Comment by jczaja** [comment]
I have checked that some ops eg. LeakyRelu grad have proto maker implemented, as grad does need X as input. So if you could point to info on implementing grad proto maker (I want all mkl-dnn activation grad ops to have X as input) that would be good.

**Comment by ghost** [comment]
Actually "X" and "Out" share same variable since activation OP is in-place edit operation.

Here is what I got from a simple "recognize_digit" model (python/paddle/fluid/tests/book/test_recognize_digit.py):
"Relu" in forward:
  ops {
    inputs {
      parameter: "X"
      arguments: "conv2d_1.tmp_1"
    }
    outputs {
      parameter: "Out"
      arguments: "conv2d_1.tmp_1"
    }
    type: "relu"
    attrs {
      name: "use_mkldnn"
      type: BOOLEAN
      b: true
    }
  }
"Relu_grad" in BWD:
  ops {
    inputs {
      parameter: "Out"
      arguments: "conv2d_1.tmp_1"
    }
    inputs {
      parameter: "Out@GRAD"
      arguments: "conv2d_1.tmp_1@GRAD"
    }
    outputs {
      parameter: "X@GRAD"
      arguments: "conv2d_1.tmp_1@GRAD"
    }
    type: "relu_grad"
    attrs {
      name: "use_mkldnn"
      type: BOOLEAN
      b: true
    }
  }

Var "conv2d_1.tmp_1" is actually both input "X" and output "Out" of relu op. Even if you get "X" as another input of relu_grad op, it will actually point to save variable as "Out".

**Comment by jczaja** [comment]
@jianhang-liu That is an interesting observation. Yes sometimes activations are inplace ops, but currently I'm guessing that not always, as inplace op is overwritting X and we may need X for grad op in its original form (for some other activations than Relu). If we include X as input to grad then
even if X and Out are pointing to the same var then no harm will be done to graph and it will work for situation when X <> Out.

**Comment by jczaja** [comment]
#18762 discuss in depth things I wanted to know and lay foundation for future manipulation of inputs for mkl-dnn activations. Hence closing.
